packages/vue/src/composables/useDenom.ts

```diff
@@ -0,0 +1,46 @@
+import { DenomTrace } from '@/utils/interfaces'
+import { Store } from 'vuex'
+
+type Response = {
+  getDenomTrace: (denom: string) => Promise<DenomTrace>
+  normalizeDenom: (denom: string) => Promise<string>
+}
+
+type Params = {
+  $s: Store<any>
+  opts?: {}
+}
+
+export default function useDenom({ $s }: Params): Response {
+  //actions
+  let queryDenomTrace = (opts: any) =>
+    $s.dispatch('ibc.applications.transfer.v1/QueryDenomTrace', opts)
+
+  // methods
+  let getDenomTrace = async (denom: string): Promise<DenomTrace> => {
+    let hash = denom.split('/')[1]
+
+    let denomTrace: DenomTrace = await queryDenomTrace({
+      options: { subscribe: false, all: false },
+      params: { hash }
+    })
+
+    return denomTrace
+  }
+  let normalizeDenom = async (denom: string): Promise<string> => {
+    let normalized = denom.toUpperCase()
+
+    let isIBC = denom.indexOf('ibc/') == 0
+
+    if (isIBC) {
+      normalized = (await getDenomTrace(denom)).denom_trace.base_denom
+    }
+
+    return normalized
+  }
+
+  return {
+    getDenomTrace,
+    normalizeDenom
+  }
```

Comment:
Two comments on this:

a) Denom traces do not change so if it's been queried before in the app session it's fine to use the fetched version.
In emeris, we handle this by mostly using the following convention:

```
const value = getter ?? await dispatch
```
(basically only dispatch the query if the getter does not return a result)

b) In emeris, we found that as the app grew and multiple components imported composables multiple times in many different places it became a performance issue. We have since switched to singleton composables. See: https://github.com/allinbits/demeris/blob/develop/src/composables/useDenoms.ts

Reply:
a) Definitely. We're going to cache this info.

b) Thanks. Good to know. Actually we should come up w/ a _composables_ package to serve our vue apps.